modules/table/src/main/java/org/apache/ignite/internal/table/distributed/raft/snapshot/incoming/IncomingSnapshotCopier.java

```diff
@@ -0,0 +1,117 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.ignite.internal.table.distributed.raft.snapshot.incoming;
+
+import java.io.IOException;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.TimeUnit;
+import org.apache.ignite.internal.table.distributed.raft.snapshot.PartitionSnapshotStorage;
+import org.apache.ignite.internal.table.distributed.raft.snapshot.SnapshotUri;
+import org.apache.ignite.internal.table.message.SnapshotMetaRequest;
+import org.apache.ignite.internal.table.message.SnapshotMetaResponse;
+import org.apache.ignite.internal.table.message.TableMessagesFactory;
+import org.apache.ignite.network.ClusterNode;
+import org.apache.ignite.network.MessagingService;
+import org.apache.ignite.network.NetworkMessage;
+import org.apache.ignite.raft.jraft.entity.RaftOutter.SnapshotMeta;
+import org.apache.ignite.raft.jraft.storage.SnapshotStorage;
+import org.apache.ignite.raft.jraft.storage.snapshot.SnapshotCopier;
+import org.apache.ignite.raft.jraft.storage.snapshot.SnapshotReader;
+
+/**
+ * Snapshot copier implementation for partitions. Used to stream partition data from the leader to the local node.
+ */
+public class IncomingSnapshotCopier extends SnapshotCopier {
+    /** Messages factory. */
+    private static final TableMessagesFactory MSG_FACTORY = new TableMessagesFactory();
+
+    /** {@link SnapshotStorage} instance for the partition. */
+    private final PartitionSnapshotStorage snapshotStorage;
+
+    /** Snapshot URI. */
+    private final SnapshotUri snapshotUri;
+
+    /** Rebalance thread-pool, used to write data into a storage. */
+    //TODO Use external pool.
+    private final ExecutorService threadPool = Executors.newSingleThreadExecutor();
+
+    /**
+     * Snapshot meta read from the leader.
+     *
+     * @see SnapshotMetaRequest
+     */
+    private SnapshotMeta snapshotMeta;
```

Comment:
Looks like this field should be `volatile`

Reply:
Technically, it shouldn't. It's being read after `join` is called

Reply:
Do you mean `join` that calls `awaitTermination`? Do you mean that you rely on `whenCompleteAsync` to be launched on the executors threads so that results of the lambda inside will be published after their termination?

Reply:
I mean that right now it's a placeholder, and "technically" (once again) I was wrong, because the implementation is sh*t, but in the future it's reasonable to expect such field without `volatile`, because `join` must give us the visibility